### PR TITLE
Fix target_train_data_fraction overriding pretrain_data_fraction

### PR DESCRIPTION
**Pull request by pyeres**
This PR addresses the bug described in issue #1066 ("Results with pretrain_data_fraction args").

This PR...
* Adds a new private field to the `Task` object (`_instance_generators`) to store instance generators with the appropriate phase-specific data fraction settings.
* Adds new methods, getter `get_instance_generator` and setter `set_instance_generator`, to `Task` to access the `_instance_generators`.
* Updates preprocessing, training, and evaluation code to use the new getter and setter.
* Updates a test that uses evaluation logic to use the new getter and setter.
* Adds tests for `Task`'s that an exception is raised when training instance generator is requested without specifying the phase (because phase determines data fraction).

Validation: the config provided in issue #1066 was run with these code changes. Results logged for the task (MNLI) using a pretraining data fraction were more in line with expectation: after 25k steps training accuracy was ~100%, while val and test accuracy were ~75%.

### CODE REVIEW

File: jiant/tasks/tasks.py [hunk 1]
```diff
@@ -293,6 +295,41 @@ def handle_preds(self, preds, batch):
...
         return preds
 
+    def set_instance_generator(
+        self, dataset_name: str, instance_generator: Iterable, phase: str = None
+    ):
+        """Takes a data instance generator and stores it in a private field of this Task instance
+
+        Parameters
+        ----------
+        dataset_name : string
```
Comment (sleepinyourhat) [inline jiant/tasks/tasks.py:305]: I'm more familiar with 'dataset' referring to a full collection, not a specific split subset of the data. How about something like 'split_name' instead?

Comment (sleepinyourhat) [reply jiant/tasks/tasks.py:305]: Otherwise, it's plausible that future code editors would try setting something like dataset_name = "mnli".

File: jiant/trainer.py [hunk 2]
```diff
@@ -322,7 +322,12 @@ def _setup_training(
                 os.mkdir(os.path.join(self._serialization_dir, task.name))
 
             # Adding task-specific smart iterator to speed up training
```
Comment (sleepinyourhat) [inline jiant/trainer.py:324]: It's okay to leave this be, but do you know what 'smart' is supposed to mean here?

Comment (sleepinyourhat) [reply jiant/trainer.py:324]: I suspect it refers to some contrast that's no longer relevant...

File: jiant/trainer.py [hunk 3]
```diff
@@ -335,7 +340,9 @@ def _setup_training(
                 biggest_batch_first=True,
             )
             task_info["iterator"] = iterator
-            task_info["tr_generator"] = iterator(task.train_data, num_epochs=None)
+            task_info["tr_generator"] = iterator(
```
Comment (sleepinyourhat) [inline jiant/trainer.py:343]: The naming seems a bit confusing here. If you're calling 'get_instance_generator', shouldn't that return something that can be used as a 'generator'?

File: jiant/trainer.py (hunk 2 above)
Comment (pyeres) [reply jiant/trainer.py]: [Looks like](https://github.com/nyu-mll/jiant/pull/87/files) it was added to describe the addition of a BucketIterator. Dropping w/ a44d4d5.

File: jiant/tasks/tasks.py (hunk 1 above)
Comment (pyeres) [reply jiant/tasks/tasks.py]: 👍, done w/ c5d6822.

File: jiant/trainer.py (hunk 3 above)
Comment (pyeres) [reply jiant/trainer.py:342]: Thanks — resolved some naming confusion in my changes w/ 72a5bfe.
